### DISCUSSION TITLE
Are there tools for generating SQLModel classes from an existing database?

### DISCUSSION THREAD

**Discussion by cfhowes** [fastapi/sqlmodel#760] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the SQLModel documentation, with the integrated search.
- [X] I already searched in Google "How to X in SQLModel" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to SQLModel but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to SQLModel but to [SQLAlchemy](https://github.com/sqlalchemy/sqlalchemy).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
pass

# i'm looking for code before i go trying to write the same code again.
```

### Description

I have an existing database (400+ tables!!!) and I would like to try and put FastAPI on top of it.  Current API is in Node JS, so I don't have python based models for it.  Are there tools to create SQLModel classes for an existing database?

if no, say I use something like https://pypi.org/project/sqlacodegen/ to create SQLAlchemy classes, is there any documentation on how to upgrade classes from plain SQLAlchemy to SQLModel?  Maybe I can write a tool more quickly then I can hand-write 400+ classes!

### Operating System

Linux

### SQLModel Version

latest.

### Python Version

3.12.1

**Comment by Baghdady92** [comment]
i suggest to use the library you suggest 
you will get sqlalchemy schema 
use chatgpt to transfer it to sqlmodel 
sqlmodel is built on top on sqlalchemy 
and 
Good Luck !

**Reply by cfhowes** [reply]
thanks for the suggestions!  I am working through some bugs/enhancements to the tools that I have found, and will let folks know how I did it once done.

**Reply by skullcrash** [reply]
In fact sqlacodegen produces sqlmodel schema too ! Look at the readme carefully :

> The following built-in generators are available:
> 
> tables (only generates Table objects, for those who don't want to use the ORM)
> declarative (the default; generates classes inheriting from declarative_base()
> dataclasses (generates dataclass-based models; v1.4+ only)
> sqlmodels (generates model classes for [SQLModel](https://sqlmodel.tiangolo.com/))

**Reply by charlie-corus** [reply]
I tried this recently, and ChatGPT complained bitterly and only half-did a single table. Maybe a more persistent "prompt engineer" could persuade it to have a better go.

`sqlcodegen` has a decent stab, but you might have to tweak the version to get it to work. 

`APILogicServer` wraps and partially reimplements `sqlcodegen`, and does a good job, but the output is very verbose.

**Comment by DurandA** [comment]
While not available yet, this is mentioned on [the roadmap](https://github.com/tiangolo/sqlmodel/issues/654):
> - [ ]  Integrated model generation from existing DBs.

**Comment by cfhowes** [comment]
Here's what I ended up doing:
 - dump schema definitions as SQL from the DB.
 - installed https://github.com/xnuinside/simple-ddl-parser (latest version)
 - used omymodels to generate the model. 
    - note that right now omymodels does not support SQLModel, but I made a branch that has worked for my case, see https://github.com/xnuinside/omymodels/issues/57 and https://github.com/cfhowes/omymodels/tree/sqlmodel_support  I'll submit it as a PR once i'm more confident it's working at least for my use case.

**Comment by nucccc** [comment]
Hi everyone,
I had this need some time ago, and I started to develop something in my free time, I called it sqlmodelgen https://github.com/nucccc/sqlmodelgen.

It mostly targets code generation from the `CREATE TABLE` sql code, but it has limited functionalities to interact with both SQLite and Postgres. I shall provide a new release in the following weeks on PyPi, at the moment I'm just trying to improve the unit tests.

It already satisfies my basic needs for work, but I'm improving it as a side project. I'm convinced it has several limitations, but I'm slowly working on it.

I don't know if this can be of any utility to anybody, but in case I hope this helps!

**Comment by Maydmor** [comment]
Yes there is [Evernox](https://evernox.com) 
You can import your Database (Postgres, MySql, Sqlite, MariaDB or BigQuery) and directly generate SQLModel classes.

Here is the Guide: 
https://www.evernox.com/generate-code-with-evernox/

**Comment by mrcMesen** [comment]
If someone comes here trying to create the SQL Models from an existing database we can use `sqlacodegen` from its [documentation](https://pypi.org/project/sqlacodegen/):

```sqlacodegen --generator sqlmodels ....rest_of_settings```
